### PR TITLE
feat(storage): supporting put api when the objectLock is enabled

### PR DESCRIPTION
**Pull request by kvramyasri7**
#### Description of changes

- add calculateMD5 utility function
- make put call successful when the ObjectLock is enabled.

#### Issue #, if available

https://github.com/aws-amplify/amplify-js/issues/8675

#### Checklist

- [X] PR description included
- [X] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/src/common/MD5utils.native.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { Md5 } from '@aws-sdk/md5-js';
+import { toBase64, fromBase64 } from '@aws-sdk/util-base64-browser';
```
Comment (AllanZhengYP) [inline packages/storage/src/common/MD5utils.native.ts:5]: We already implemented the base64 encoding, so no need for this dependency: https://github.com/aws-amplify/amplify-js/tree/main/packages/storage/src/AwsClients/S3/runtime/base64

Comment (kvramyasri7) [reply packages/storage/src/common/MD5utils.native.ts]: Thanks. Just used the existing implementation.

File: packages/storage/src/AwsClients/S3/putObject.ts [hunk 2]
```diff
@@ -55,10 +57,13 @@ const putObjectSerializer = async (
...
-	const headers = await serializeObjectConfigsToHeaders({
-		...input,
-		ContentType: input.ContentType ?? 'application/octet-stream',
-	});
+	const headers = {
+		...(await serializeObjectConfigsToHeaders({
+			...input,
+			ContentType: input.ContentType ?? 'application/octet-stream',
+		})),
+		...assignStringVariables({ 'content-md5': input.ContentMD5 }),
```
Comment (cshfang) [inline packages/storage/src/AwsClients/S3/putObject.ts:65]: Instead of spreading the result of `serializeObjectConfigsToHeaders` to build yet another new object, would it make more sense (or event possible) to add `assignStringVariables` into the `serializeObjectConfigsToHeaders` function?

File: packages/storage/src/AwsClients/S3/uploadPart.ts [hunk 3]
```diff
@@ -43,7 +45,10 @@ const uploadPartSerializer = async (
 	input: UploadPartInput,
 	endpoint: Endpoint
 ): Promise<HttpRequest> => {
-	const headers = await serializeObjectSsecOptionsToHeaders(input);
+	const headers = {
```
Comment (cshfang) [inline packages/storage/src/AwsClients/S3/uploadPart.ts:48]: Same question

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts [hunk 4]
```diff
@@ -67,6 +71,17 @@ export class AWSS3ProviderManagedUpload {
 
 	public async upload() {
 		try {
+			const { isObjectLockEnabled }: { isObjectLockEnabled: boolean } =
+				this.opts;
```
Comment (cshfang) [inline packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:75]: Seems like a useless typing here. Possible to try to type `opts` a bit (i.e. to include this property)?

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts [hunk 5]
```diff
@@ -67,6 +71,17 @@ export class AWSS3ProviderManagedUpload {
...
+			if (typeof isObjectLockEnabled !== 'boolean') {
+				logger.error('isObjectLockEnabled can be either true/false');
+				throw Error('isObjectLockEnabled can be either true/false');
```
Comment (cshfang) [inline packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:78]: Nit: `isObjectLockEnabled must be a boolean value`

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts [hunk 6]
```diff
@@ -67,6 +71,17 @@ export class AWSS3ProviderManagedUpload {
 
 	public async upload() {
 		try {
+			const { isObjectLockEnabled }: { isObjectLockEnabled: boolean } =
+				this.opts;
+			if (typeof isObjectLockEnabled !== 'boolean') {
```
Comment (cshfang) [inline packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:76]: Can `opts` be typed somewhere to have an optional `isObjectLockEnabled` property? In theory, we should be enforcing this via typing instead, right? We don't really do this kind of runtime checking everywhere (or it'd be endless) so I'm curious what prompted this?

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts [hunk 7]
```diff
@@ -67,6 +71,17 @@ export class AWSS3ProviderManagedUpload {
...
+				logger.error('isObjectLockEnabled can be either true/false');
+				throw Error('isObjectLockEnabled can be either true/false');
+			}
+			if (isObjectLockEnabled) {
+				this.params.ContentMD5 = await calculateContentMd5(
+					this.params.Body as Blob | File | string
```
Comment (cshfang) [inline packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:82]: Nit: Worth extracting `Blob | File | string` as a type?

Comment (AllanZhengYP) [inline packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:82]: Instead of casting to union type(which doesn't make a lot sense), can you use `@ts-expect-error` comment? ([link](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#-ts-expect-error-comments)) and a description of why the does does not match and why it's ok to use.

Comment (cshfang) [reply packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:82]: Also curious how this is satisfying the `.native` version of this...
Passing a `A | B | C` type to a function expecting `A | B` shouldn't work (unless there are overlaps between the types but, in this case, I don't believe there is..)

File: packages/storage/src/common/MD5utils.native.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+	if (typeof content === 'string') {
+		hasher.update(content);
+	} else {
+		const buffer = await readFileToBase64(content);
+		hasher.update(utf8Encode(buffer));
+	}
+	const digest = await hasher.digest();
+	return toBase64(digest);
+};
+const readFileToBase64 = (blob: Blob): Promise<string> => {
```
Comment (AllanZhengYP) [inline packages/storage/src/common/MD5utils.native.ts:20]: nit: empty line
```suggestion

const readFileToBase64 = (blob: Blob): Promise<string> => {
```

File: packages/storage/src/AwsClients/S3/putObject.ts (hunk 2 above)
Comment (AllanZhengYP) [reply packages/storage/src/AwsClients/S3/putObject.ts:65]: The `serializeObjectConfigsToHeaders` function is used in more APIs than `content-md5` is needed. Like create multipart upload API. Here we can limit the effect of adding md5 hash to only modified APIs.

File: packages/storage/src/common/MD5utils.native.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+	return new Promise((resolve, reject) => {
+		const reader = new FileReader();
+		reader.onloadend = () => {
+			// reference: https://developer.mozilla.org/en-US/docs/Web/API/FileReader/readAsDataURL
+			// response from readAsDataURL is always prepended with "data:*/*;base64,"
+			if (reader.readyState !== 2) {
+				return reject(new Error('Reader aborted too early'));
+			}
+			const result = reader.result as string;
+			const dataOffset = result.indexOf(',') + 1;
```
Comment (cshfang) [inline packages/storage/src/common/MD5utils.native.ts:30]: Is this comma related to the comment on ln24/25? If it's always prepended with the same thing, can't we just substring at the same place everytime without this look up?

File: packages/storage/src/common/MD5utils.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+};
+
+const readFile = (file: File | Blob): Promise<ArrayBuffer> => {
+	return new Promise((resolve, reject) => {
+		const reader = new FileReader();
+		reader.onloadend = () => {
+			if (reader.result) {
+				resolve(reader.result as ArrayBuffer);
+			} else {
+				reject(new Error('Failed to read file!'));
```
Comment (cshfang) [inline packages/storage/src/common/MD5utils.ts:28]: Nit:
```suggestion
				reject(new Error('Failed to read file.'));
```

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts (hunk 7 above)
Comment (cshfang) [reply packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:82]: Actually if File is a subclass of Blob then should be ok

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts (hunk 4 above)
Comment (kvramyasri7) [reply packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:75]: Yeah. We can remove this typing here 👍.

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts (hunk 6 above)
Comment (kvramyasri7) [reply packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:75]: This check is to avoid calling `calculateContentMd5` when the isObjectLockEnabled is set but not `true`. Slightly changing the logic. what do you think?

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts (hunk 7 above)
Comment (kvramyasri7) [reply packages/storage/src/providers/AWSS3ProviderManagedUpload.ts]: Thanks for the suggestion. Updated.

Comment (kvramyasri7) [reply packages/storage/src/providers/AWSS3ProviderManagedUpload.ts]: used @ts-expect-error. what do you think?

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts [hunk 11]
```diff
@@ -67,6 +71,19 @@ export class AWSS3ProviderManagedUpload {
...
+					this.params.ContentMD5 = await calculateContentMd5(
+						// @ts-expect-error currently ReadableStream<any> is not being supported in put api
+						this.params.Body
+					);
+				} else {
+					// Should not call calculateContentMd5 when the isObjectLockEnabled is set but not true
+					logger.error('isObjectLockEnabled must be a boolean value');
+					throw Error('isObjectLockEnabled must be a boolean value');
+				}
+			}
```
Comment (AllanZhengYP) [inline packages/storage/src/providers/AWSS3ProviderManagedUpload.ts:86]: You can get less if statements with:

```suggestion
			if (isObjectLockEnabled != null && typeof isObjectLockEnabled !== 'boolean') {
			  	// Should not call calculateContentMd5 when the isObjectLockEnabled is set but not true
					logger.error('isObjectLockEnabled must be a boolean value');
					throw Error('isObjectLockEnabled must be a boolean value');
			}
		  if (isObjectLockEnabled) {
					this.params.ContentMD5 = await calculateContentMd5(
						  // @ts-expect-error currently ReadableStream<any> is not being supported in put api
						  this.params.Body
					);
				} 
```

File: packages/storage/src/common/MD5utils.native.ts (hunk 9 above)
Comment (kvramyasri7) [reply packages/storage/src/common/MD5utils.native.ts]: Just slightly changed the code. Please let me know if that works.

File: packages/storage/src/common/MD5utils.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+		hasher.update(content);
+	} else {
+		const buffer = await readFile(content);
+		hasher.update(buffer);
+	}
+	const digest = await hasher.digest();
+	return toBase64(digest);
+};
+
+const readFile = (file: File | Blob): Promise<ArrayBuffer> => {
```
Comment (AllanZhengYP) [inline packages/storage/src/common/MD5utils.ts:21]: `Blob` only is enough here. I believe `File` is a subclass of `Blob`

File: packages/storage/src/common/MD5utils.ts [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+	return new Promise((resolve, reject) => {
+		const reader = new FileReader();
+		reader.onloadend = () => {
+			if (reader.result) {
+				resolve(reader.result as ArrayBuffer);
+			} else {
+				reject(new Error('Failed to read file.'));
+			}
+		};
```
Comment (AllanZhengYP) [inline packages/storage/src/common/MD5utils.ts:30]: Missing `onerror` callback

**Review by AllanZhengYP — COMMENTED** [review]
It looks good. I have 2 more comments.

File: packages/storage/src/common/MD5utils.ts (hunk 13 above)
Comment (kvramyasri7) [reply packages/storage/src/common/MD5utils.ts:30]: Updated. Thanks.

File: packages/storage/src/providers/AWSS3ProviderManagedUpload.ts [hunk 14]
```diff
@@ -67,6 +71,18 @@ export class AWSS3ProviderManagedUpload {
 
 	public async upload() {
 		try {
+			const { isObjectLockEnabled } = this.opts;
+			if (isObjectLockEnabled != null && typeof isObjectLockEnabled !== 'boolean') {
+			  	// Should not call calculateContentMd5 when the isObjectLockEnabled is set but not true
+					logger.error('isObjectLockEnabled must be a boolean value');
+					throw Error('isObjectLockEnabled must be a boolean value');
+			}
+		  if (isObjectLockEnabled) {
```
[the remark this answers is not in the document]
Comment (kvramyasri7) [reply packages/storage/src/providers/AWSS3ProviderManagedUpload.ts]: this seems outdated 🤔

File: packages/storage/src/common/MD5utils.native.ts [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+	return toBase64(digest);
+};
+
+const readFileToBase64 = (blob: Blob): Promise<string> => {
+	return new Promise((resolve, reject) => {
+		const reader = new FileReader();
+		reader.onloadend = () => {
+			// reference: https://developer.mozilla.org/en-US/docs/Web/API/FileReader/readAsDataURL
+			// response from readAsDataURL is always prepended with "data:*/*;base64,"
+			if (reader.readyState !== 2) {
```
[the remark this answers is not in the document]
Comment (kvramyasri7) [reply packages/storage/src/common/MD5utils.native.ts:28]: Added thanks for the suggestion.

**Review by cshfang — APPROVED** [review]

**Review by AllanZhengYP — APPROVED** [review]
🚢🚢🚢
